src/_posts/2017-08-04-splittable-do-fn.md

```diff
@@ -0,0 +1,449 @@
+which reads the results of a `PCollection` of queries from Cloud Spanner,
+compared to
+[`SpannerIO.read()`](https://github.com/apache/beam/blob/f7e8f886c91ea9d0b51e00331eeb4484e2f6e000/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java#L318)
+which executes a single query. We believe such dynamic data sources are a very
+useful capability, often overlooked by other data processing frameworks.
+
+## When ParDo and GroupByKey are not enough
+
+However, in some cases, `ParDo`, `GroupByKey` and their derivatives, despite their
+flexibility, are insufficient to build an efficient IO connector.
+
+For example, imagine reading files using the sequence `ParDo(filepattern →
+expand into files)`, `ParDo(filename → read records)`, or reading a Kafka topic
+using `ParDo(topic → list partitions)`, `ParDo(topic, partition → read
+records)`. This approach has two big issues:
+
+* In the file example, some files might be much larger than others, so the
+second `ParDo` may have very long individual `@ProcessElement` calls, and the
+pipeline can suffer from poor performance due to stragglers.
+
+* In the Kafka example, implementing the second `ParDo` is *simply impossible*
+with a regular `DoFn`, because it would need to output an infinite number of
+records per each input element `topic, partition`. *([stateful
+processing]({{ site.baseurl }}/blog/2017/02/13/stateful-processing.html)
+comes close, but it has other limitations that make it insufficient for this
+task*).
+
+## Beam Source API
+
+Apache Beam historically provides a Source API
+([BoundedSource]({{ site.baseurl }}/documentation/sdks/javadoc/2.0.0/org/apache/beam/sdk/io/BoundedSource.html)
+and
+[UnboundedSource]({{ site.baseurl }}/documentation/sdks/javadoc/2.0.0/org/apache/beam/sdk/io/UnboundedSource.html))
+which does not have these limitations and allows developing efficient data
+sources for batch and streaming systems.  This API can be used in a pipeline
+using the
+[`Read.from(Source)`]({{ site.baseurl }}/documentation/sdks/javadoc/2.0.0/org/apache/beam/sdk/io/Read.html)
+built-in `PTransform`.
+
+The Source API is largely similar to that of most other data processing
+frameworks, and allows the system to read data in parallel using multiple
+workers, as well as checkpoint and resume reading from an unbounded data source.
+Additionally, the Beam
+[`BoundedSource`]({{ site.baseurl }}/documentation/sdks/javadoc/2.0.0/org/apache/beam/sdk/io/BoundedSource.html)
+API provides advanced features such as progress reporting and [dynamic
+rebalancing]({{ site.baseurl }}/blog/2016/05/18/splitAtFraction-method.html)
+(which together enable autoscaling), and
+[`UnboundedSource`]({{ site.baseurl }}/documentation/sdks/javadoc/2.0.0/org/apache/beam/sdk/io/UnboundedSource.html)
+supports reporting the source's watermark and backlog.
+
```

Comment:
Is it worth also calling out the non-unification between bounded and unbounded pipelines?